src/shm/pool/slot.rs

```diff
@@ -0,0 +1,412 @@
+        let mut free = self.free_list.lock().unwrap();
+        if let Some((off, len)) = free.last_mut() {
+            if *off + *len == old_len {
+                *len += new_len - old_len;
+                return Ok(());
+            }
+        }
+        free.push((old_len, new_len - old_len));
+        Ok(())
+    }
+
+    fn alloc(&mut self, size: usize) -> io::Result<usize> {
+        let mut free = self.free_list.lock().unwrap();
+        for (offset, len) in free.iter_mut() {
+            if *len >= size {
+                let rv = *offset;
+                *len -= size;
+                *offset += size;
+                return Ok(rv);
+            }
+        }
+        let mut rv = self.inner.len();
+        let mut pop_tail = false;
+        if let Some((start, len)) = free.last() {
+            if start + len == self.inner.len() {
+                rv -= len;
+                pop_tail = true;
+            }
+        }
+        // resize like Vec::reserve, always at least doubling
+        let target = std::cmp::max(rv + size, self.inner.len() * 2);
+        self.inner.resize(target)?;
+        // adjust the end of the freelist here
+        if pop_tail {
+            free.pop();
+        }
+        if target > rv + size {
+            free.push((rv + size, target - rv - size));
+        }
+        Ok(rv)
+    }
+
+    fn free(free_list: &Mutex<Vec<(usize, usize)>>, mut offset: usize, mut len: usize) {
+        let mut free = free_list.lock().unwrap();
+        let mut nf = Vec::with_capacity(free.len() + 1);
+        for &(ioff, ilen) in free.iter() {
+            if ioff + ilen == offset {
+                offset = ioff;
+                len += ilen;
+                continue;
+            }
+            if ioff == offset + len {
+                len += ilen;
+                continue;
+            }
+            if ioff > offset + len && len != 0 {
+                nf.push((offset, len));
+                len = 0;
+            }
+            if ilen != 0 {
+                nf.push((ioff, ilen));
+            }
+        }
+        if len != 0 {
+            nf.push((offset, len));
+        }
+        *free = nf;
+    }
+
+    /// Create a new slot with the given size in bytes.
+    pub fn new_slot(&mut self, len: usize) -> io::Result<Slot> {
+        let len = (len + 15) & !15;
+        let offset = self.alloc(len)?;
+
+        Ok(Slot {
+            inner: Arc::new(SlotInner {
+                free_list: Arc::downgrade(&self.free_list),
+                offset,
+                len,
+                active_buffers: AtomicUsize::new(0),
+            }),
+        })
+    }
+
+    /// Get the bytes corresponding to a given slot.
+    ///
+    /// Note: prefer using canvas, which will prevent drawing to a buffer that has not been
+    /// released by the server.
+    ///
+    /// Returns an empty buffer if the slot does not belong to this pool.
+    pub fn raw_data_mut(&mut self, slot: &Slot) -> &mut [u8] {
```

Comment:
I assume this is for higher level pools based on the slot pool?

Reply:
Yes, or for things like using read-only access to a buffer that is in use by the server.